Class/Practise 2.1/tests.py

```diff
@@ -0,0 +1,100 @@
+import subprocess
+import sys
+import os
+
+def run_program(program_path, *args):
+    result = subprocess.run([program_path, *args], capture_output=True, text=True)
+    return result.stdout.strip(), result.stderr.strip()
+
+def test_case_1(program_path):
+    # Step 1: Ensure fresh start
+    if os.path.exists("user_stats.txt"):
+        os.remove("user_stats.txt")
+    
+    # Step 2: Run with "Alice"
+    output, error = run_program(program_path, "Alice")
+    assert output == "Welcome, Alice!"
+    
+    # Step 3: Run again with "Alice"
+    output, error = run_program(program_path, "Alice")
+    assert output == "Hello again(x2), Alice!"
+    print("Test Case 1 Passed")
+
+def test_case_2(program_path):
+    # Step 1: Run with "Bob"
+    run_program(program_path, "Bob")
+    
+    # Step 2: Run again with "Bob"
+    run_program(program_path, "Bob")
+    
+    # Step 3: Reset statistics
+    output, error = run_program(program_path, "Bob", "delete")
+    assert output == "Statistics for Bob have been reset."
+    
+    # Step 4: Run again with "Bob"
+    output, error = run_program(program_path, "Bob")
+    assert output == "Welcome, Bob!"
+    print("Test Case 2 Passed")
+
+def test_case_3(program_path):
+    # Step 1: Run with "Charlie"
+    run_program(program_path, "Charlie")
+    
+    # Step 2: Run with "Dana"
+    run_program(program_path, "Dana")
```

Comment:
I didn't assert for the same reason as in the test case 2 because the this test case is aimed at testing "bread" functionality so there is no need to do extra assertions. 
However, tell me if I am wrong I can add them

Reply:
ok. It is up to you and your intentions.

Some general idea for coverage extension: What can be done to improve such type of tests is for example generation or reuse of some predefined database of file for the test. For example if we are not interested for multiple calls of program as a part of test then we can just emulate it by predefined file.
Also pregenerated file can allow you to test more scenarios like some program behavior after 1000 of program calls - it can not be a manual step but it can be emulated.

Reply:
Yeah, I also thought about creating seperate file for testing, however, I found that for this small program several call of will be fine as they are lightweight.

Reply:
It is a regular situation to have supplementary datasets for test cases - it is a part of test precondition. Therefore you can use that approach in further tasks if needed